crates/cxx-qt-lib/src/types/qstring.rs

```diff
@@ -3,52 +3,105 @@
+        fn qstringInitFromRustString(string: &str) -> QString;
         #[namespace = "rust::cxxqtlib1"]
         #[rust_name = "qstring_init_from_qstring"]
-        fn qstringInitFromQString(string: &QString) -> UniquePtr<QString>;
-    }
+        fn qstringInitFromQString(string: &QString) -> QString;
 
-    impl UniquePtr<QString> {}
+        #[namespace = "rust::cxxqtlib1"]
+        #[rust_name = "qstring_to_rust_string"]
+        fn qstringToRustString(string: &QString) -> String;
+    }
 }
 
 /// The QString class provides a Unicode character string.
 ///
-/// Note that this is the C++ representation and String or &str should be used in Rust.
-pub type QString = ffi::QString;
+/// Note that QString is a UTF-16 whereas Rust strings are a UTF-8
+pub struct QString {
+    /// The layout has changed between Qt 5 and Qt 6
+    #[cfg(qt_version_major = "5")]
+    _space: MaybeUninit<usize>,
+    #[cfg(qt_version_major = "6")]
+    _space: MaybeUninit<[usize; 2]>,
+}
 
-impl QString {
-    pub fn from_ref(string: &QString) -> cxx::UniquePtr<Self> {
-        ffi::qstring_init_from_qstring(string)
+impl Clone for QString {
+    /// Constructs a copy of other.
+    ///
+    /// This operation takes constant time, because QString is implicitly shared.
+    /// This makes returning a QString from a function very fast.
+    /// If a shared instance is modified, it will be copied (copy-on-write), and that takes linear time.
+    fn clone(&self) -> Self {
+        ffi::qstring_init_from_qstring(self)
     }
+}
 
-    pub fn from_str(str: &str) -> cxx::UniquePtr<Self> {
-        ffi::qstring_init_from_rust_string(str)
+impl Default for QString {
+    /// Constructs an invalid color. An invalid color is a color that is not properly set up for the underlying window system.
+    ///
+    /// The alpha value of an invalid color is unspecified.
+    fn default() -> Self {
+        ffi::qstring_init_default()
     }
 }
 
 impl std::fmt::Display for QString {
+    /// Convert the QString to a Rust string
+    ///
+    /// Note that this performs a UTF conversion
     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
         write!(f, "{}", ffi::qstring_to_rust_string(self))
     }
 }
 
-impl From<&QString> for cxx::UniquePtr<QString> {
-    fn from(value: &QString) -> cxx::UniquePtr<QString> {
-        QString::from_ref(value)
+impl Drop for QString {
+    /// Destroys the string.
+    fn drop(&mut self) {
+        ffi::qstring_drop(self)
     }
 }
+
+impl From<&str> for QString {
+    /// Constructs a QString from a Rust string
+    ///
+    /// Note that this performs a UTF conversion
+    fn from(str: &str) -> Self {
+        ffi::qstring_init_from_rust_string(str)
+    }
+}
+
+impl From<&String> for QString {
+    /// Constructs a QString from a Rust string
+    ///
+    /// Note that this performs a UTF conversion
+    fn from(str: &String) -> Self {
+        ffi::qstring_init_from_rust_string(str)
+    }
+}
```

Comment:
Could this redundant impl be avoided by using `impl From<AsRef<str>>`?

Reply:
Ah seems it doesn't work when i compile it further, it then wants AsRef to be dyn and requires the T to be Sized